// ng-search-filter/esm2015/lib/ng-search-filter.pipe.js
import { Pipe, Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./ng-search-filter.service";
export class NgSearchFilterPipe {
    constructor(_ngSearchFilterService) {
        this._ngSearchFilterService = _ngSearchFilterService;
        this._lang = this._ngSearchFilterService.getDefaultLang();
    }
    /**
     * @param items object from array
     * @param term term's search
     * @param excludes array of strings which will ignored during search
     */
    transform(items, term, excludes = []) {
        if (!term || !items)
            return items;
        return NgSearchFilterPipe.filter(items, term, excludes, this._lang);
    }
    /**
     *
     * @param items List of items to filter
     * @param term  a string term to compare with every property of the list
     * @param excludes List of keys which will be ignored during search
     *
     */
    static filter(items, term, excludes, lang) {
        const toCompare = term.toLocaleLowerCase(lang);
        function checkInside(item, term) {
            if (typeof item === 'string' &&
                item.toString().toLocaleLowerCase(lang).includes(toCompare)) {
                return true;
            }
            for (let property in item) {
                if (item[property] === null ||
                    item[property] == undefined ||
                    excludes.includes(property)) {
                    continue;
                }
                if (typeof item[property] === 'object') {
                    if (checkInside(item[property], term)) {
                        return true;
                    }
                }
                else if (item[property].toString().toLocaleLowerCase(lang).includes(toCompare)) {
                    return true;
                }
            }
            return false;
        }
        return items.filter(function (item) {
            return checkInside(item, term);
        });
    }
}
NgSearchFilterPipe.ɵfac = function NgSearchFilterPipe_Factory(t) { return new (t || NgSearchFilterPipe)(i0.ɵɵdirectiveInject(i1.NgSearchFilterService)); };
NgSearchFilterPipe.ɵpipe = i0.ɵɵdefinePipe({ name: "filter", type: NgSearchFilterPipe, pure: false });
NgSearchFilterPipe.ɵprov = i0.ɵɵdefineInjectable({ token: NgSearchFilterPipe, factory: NgSearchFilterPipe.ɵfac });
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(NgSearchFilterPipe, [{
        type: Pipe,
        args: [{
                name: 'filter',
                pure: false,
            }]
    }, {
        type: Injectable
    }], function () { return [{ type: i1.NgSearchFilterService }]; }, null); })();
